### DISCUSSION TITLE
Optimize notification layout

### DISCUSSION THREAD

**Discussion by jmlich** [InfiniTimeOrg/InfiniTime#1880] [Ideas]
If I understand correctly, notification consist of sender[32], title[128], message[128], category[1], icon[1]
(sizes are taken from https://github.com/piggz/harbour-amazfish/blob/11a9ea5e52e8afb697f3766fd1b7ad98095d72aa/daemon/src/services/alertnotificationservice.cpp#L38 )

btw 128 looks like bug, because I have seen only 100 (there is trim to 230 in the end) , but InfinityTime is for me hard to read.

The page it self looks like:
"Notifications" [id]/[count]\n
[sender]\n
[title]\n
[message]\n

We have 7 lines and 20 leters per each available.

In my case is the sender most often "teleports.ubports_teleports", which consumes 2 lines, then is title which contains name of person and \n and 4 lines left for message it self. 

It would be great to utilize "category" and icon for that.
https://github.com/piggz/harbour-amazfish/blob/11a9ea5e52e8afb697f3766fd1b7ad98095d72aa/daemon/src/services/alertnotificationservice.h#L94
I would like to see something like that
[telegram icon] [telegram] [id]/[count]\n
[title]\n
[message]\n

Word notification can be replaced according to application category or with sender. Icon would make orientation faster.

It looks like amazfish should trim title, because it seems it contain \n

Not sure which are troubles in infinitime and which are troubles of amazfish.

**Comment by FintasticMan** [comment]
As far as InfiniTime is concerned, notifications are strings 100 characters long, with a null terminator between the title and message body. The companion app can choose how it structures the message beyond that.

For icons, we would need to store those icons in a font, but given how memory and flash limited we are, we aren't going to do that.

Notifications have a category, which can be one of the following: Unknown, SimpleAlert, Email, News, IncomingCall, MissedCall, Sms, VoiceMail, Schedule, HighProriotyAlert, InstantMessage. At the moment, InfiniTime doesn't do much with the category, other than displaying incoming calls as calls. There is an open pull request that displays the category when viewing the notification.

**Reply by jmlich** [reply]
https://github.com/InfiniTimeOrg/InfiniTime/pull/1697

**Reply by jmlich** [reply]
I see the issue. The harbour-amazfish started to add 0x0a in https://github.com/piggz/harbour-amazfish/commit/0638cc79c93682ea88721bf402324995775ba48f (February 2021). The Infinitime started to parse \0 in
https://github.com/InfiniTimeOrg/InfiniTime/commit/03de1c67393fcb99b5987514be6f470349c57c0f (April 2021).

I have made pull request https://github.com/piggz/harbour-amazfish/pull/275

**Reply by jmlich** [reply]
By the way, where does the 100-byte limitation come from? It seems that ANS defines only 18 bytes for the entire message. It would be greatly appreciated to have more characters available, as messages are often truncated even when there's space for two additional lines on the screen

**Reply by FintasticMan** [reply]
We created the 100 char limit, so that notifications only ever take up 505 bytes in total. This is because we often have issues with too much memory being used, so it's nice to know that notifications can't take up too much.

**Comment by jmlich** [comment]
I have updated the way how title is composed in Amazfish and now I am more less happy with its layout.
https://github.com/piggz/libwatchfish/pull/4
Thanks for valuable feedback
